### ISSUE TITLE
perf(feature): MKDDescriptor spends 75% of its time in a broadcast `emb1 * emb2` + `sum(dim=(2,3))` in ExplicitSpacialEncoding — 1.4 s for 2048 patches on CPU and MPS alike

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4534]
## Problem

`MKDDescriptor(32)` on 2048 32×32 patches takes **1.41 s on the M1 CPU and 1.43 s on MPS** (autograd on), 0.43 s under `no_grad` on CPU; `SIFTDescriptor(32)` on the same patches takes 0.34 s / 0.034 s. At 16384 patches the MPS run asks for a 10.9 GiB buffer. cProfile puts 0.32 s of the 0.41 s no-grad forward in `ExplicitSpacialEncoding.forward` (`kornia/feature/mkd.py:375–398`):

```python
emb1 = torch.index_select(x, 1, idx1)     # (B, K, H, W) with K = |kron| pairs
output = emb1 * self.emb2                 # broadcast product, (B, K, H, W)
output = output.sum(dim=(2, 3))           # (B, K)
```

The dispatch-level breakdown of one forward is 784 ms in 11 `mul` calls, 155 ms in 2 `index_select`, 151 ms in 4 `sum.dim_IntList` — i.e. materialising the `(B, K, H, W)` product and reducing it, twice (two encodings in the `Sequential`). It is a contraction: `output[b, k] = sum_hw x[b, idx1[k], h, w] * emb2[k, h, w]`.

## Fix

Write it as a matmul: `x.flatten(2) @ emb2.flatten(1).T`-style with the `kron` gather folded into the weight (`W[k, c, hw] = emb2[k, hw] if c == idx1[k] else 0` is sparse, but grouping `k` by `idx1[k]` gives one `bmm` per input channel, or a single `einsum("bchw,kchw->bk")` over a `(K, C, H, W)` weight that is zero off the selected channel). Either removes the `(B, K, H, W)` intermediate; `sum(dim=(2,3))` should also go (it is a trailing-two-dims reduction, a slow path on MPS, #4522).

Measured on torch nightly 2.15.0.dev20260913, Apple M1 8 GB, macOS 26.5.1, kornia `633951bd7`, `PYTORCH_ENABLE_MPS_FALLBACK` unset, `torch.utils.benchmark` medians with `mps.synchronize` inside the timed region, fresh process per op. Split out of the MPS-audit tracking issue #4521; the PyTorch-side kernel findings are in #4522.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
